csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs

```diff
@@ -224,48 +272,90 @@ static void BuildNamedTypeId(this INamedTypeSymbol named, Context cx, TextWriter
-            else if (IsReallyUnbound(named))
+            }
+            else if (named.IsReallyUnbound())
             {
+                AddContaining();
                 trapFile.Write(named.Name);
                 trapFile.Write("`");
                 trapFile.Write(named.TypeParameters.Length);
+                // Some types such as `<>f__AnonymousType0` are not considered anonymous types by Roslyn,
+                // perhaps because they contain type parameters. We still need to treat them like anonymous
+                // types, though, by adding the underlying properties, in order to disambiguate them
+                if (named.Name.Contains("__AnonymousType"))
+                {
+                    trapFile.Write('<');
+                    trapFile.BuildList(",", named.GetMembers().OfType<IPropertySymbol>(), (prop, tb0) =>
+                        {
+                            tb0.Write(prop.Name);
+                            tb0.Write(" ");
+                            prop.Type.BuildOrWriteId(cx, tb0, symbolBeingDefined, addBaseClass);
+                        });
+                    trapFile.Write('>');
+                }
             }
             else
             {
-                subTermAction(cx, trapFile, named.ConstructedFrom);
+                named.ConstructedFrom.BuildOrWriteId(cx, trapFile, symbolBeingDefined, addBaseClass);
                 trapFile.Write('<');
                 // Encode the nullability of the type arguments in the label.
                 // Type arguments with different nullability can result in 
                 // a constructed type with different nullability of its members and methods,
                 // so we need to create a distinct database entity for it.
                 trapFile.BuildList(",", named.GetAnnotatedTypeArguments(),
-                    (ta, tb0) => subTermAction(cx, tb0, ta.Symbol)
+                    (ta, tb0) => ta.Symbol.BuildOrWriteId(cx, tb0, symbolBeingDefined, addBaseClass)
                     );
                 trapFile.Write('>');
             }
+
+            if (addBaseClass && named.GetNonObjectBaseType(cx) is INamedTypeSymbol @base)
+            {
+                // We need to limit unfolding of base classes. For example, in
+                //
+                // ```csharp
+                // class C1<T> { }
+                // class C2<T> : C1<C3<T>> { }
+                // class C3<T> : C1<C2<T>> { }
+                // class C4 : C3<C4> { }
+                // ```
+                //
+                // when we generate the label for `C4`, the base class `C3<C4>` has itself `C1<C2<C4>>` as
+                // a base class, which in turn has `C1<C3<C4>>` as a base class. The latter has the original
+                // base class `C3<C4>` as a type argument, which would lead to infinite unfolding.
+                trapFile.Write(" : ");
+                @base.BuildOrWriteId(cx, trapFile, symbolBeingDefined, false);
```

Comment:
I would add the parameter name to this call, just to make it even more explicit what that `false` means. `addBaseClass: false`